natsyncd/natsync.cpp

```diff
@@ -0,0 +1,1047 @@
+ *       or dynamic Twice NAT entry, then it is a Twice NAPT entry.
+ */
+int NatSync::addNatEntry(struct nfnl_ct *ct, struct naptEntry &entry, bool addFlag)
+{
+    SWSS_LOG_ENTER();
+
+    bool src_ip_natted   = (entry.orig_src_ip      != entry.nat_src_ip);
+    bool dst_ip_natted   = (entry.orig_dest_ip     != entry.nat_dest_ip);
+    bool src_port_natted = (src_ip_natted && ((entry.orig_src_l4_port != entry.nat_src_l4_port) || 
+                                              (matchingSnaptPoolExists(entry.nat_src_ip)) || 
+                                              (matchingSnaptEntryExists(entry))));
+    bool dst_port_natted = (dst_ip_natted && ((entry.orig_dst_l4_port != entry.nat_dst_l4_port) ||
+                                              (matchingDnaptEntryExists(entry))));
+    bool entryExists = 0, reverseEntryExists = 0;
+
+    SWSS_LOG_INFO("Flags: src natted %d, dst natted %d, src port natted %d, dst port natted %d", src_ip_natted, dst_ip_natted, src_port_natted, dst_port_natted);
+
+    std::vector<FieldValueTuple> fvVector, reverseFvVector;
+    string protostr = ((entry.protocol == IPPROTO_TCP) ? "TCP:" : "UDP:");
+    string opStr    = ((addFlag) ? "CREATE" : "DELETE");
+    string key = "", reverseEntryKey = "";
+
+    FieldValueTuple snat_type("nat_type", "snat");
+    FieldValueTuple dnat_type("nat_type", "dnat");
+    FieldValueTuple dynamic_entry("entry_type", "dynamic");
+
+    if (src_ip_natted && dst_ip_natted)
+    {
+        if (addFlag)
+        {
+            FieldValueTuple translated_src_ip("translated_src_ip", entry.nat_src_ip.to_string());
+            FieldValueTuple translated_dst_ip("translated_dst_ip", entry.nat_dest_ip.to_string());
+            FieldValueTuple reverse_translated_src_ip("translated_src_ip", entry.orig_dest_ip.to_string());
+            FieldValueTuple reverse_translated_dst_ip("translated_dst_ip", entry.orig_src_ip.to_string());
+
+            fvVector.push_back(dynamic_entry);
+            fvVector.push_back(translated_src_ip);
+            fvVector.push_back(translated_dst_ip);
+
+            reverseFvVector.push_back(dynamic_entry);
+            reverseFvVector.push_back(reverse_translated_src_ip);
+            reverseFvVector.push_back(reverse_translated_dst_ip);
+        }
+
+        string tmpKey             = key + entry.orig_src_ip.to_string() + ":" + entry.orig_dest_ip.to_string();
+        string tmpReverseEntryKey = reverseEntryKey + entry.nat_dest_ip.to_string() + ":" + entry.nat_src_ip.to_string();
+
+        std::vector<FieldValueTuple> values;
+        if (m_twiceNatCheckTable.get(tmpKey, values))
+        {
+            src_port_natted = dst_port_natted = false;
+
+            for (auto iter : values)
+            {
+                /* If a matching Static Twice NAT entry exists in the APP_DB,
+                 * it has higher priority than the dynamic twice nat entry. */
+                if ((fvField(iter) == "entry_type") && (fvValue(iter) == "static"))
```

Comment:
For static snat entry, the packet can come up in case of h/w L3 destination lookup failure. To handle such cases, we need the SNAT rule for static entry in the iptables.